File: sceneview/src/main/java/com/google/ar/sceneform/rendering/ResourceManager.java
package com.google.ar.sceneform.rendering;

import androidx.annotation.Nullable;

import com.google.ar.sceneform.resources.ResourceHolder;
import com.google.ar.sceneform.resources.ResourceRegistry;

import java.util.ArrayList;

/**
 * Minimal resource manager. Maintains mappings from ids to created resources and a task executor
 * dedicated to loading resources asynchronously.
 *
 * @hide
 */
@SuppressWarnings("initialization") // Suppress @UnderInitialization warning.
public class ResourceManager {
    @Nullable
    private static ResourceManager instance = null;

    private final ArrayList<ResourceHolder> resourceHolders = new ArrayList<>();
    private final ResourceRegistry<Texture> textureRegistry = new ResourceRegistry<>();
    private final ResourceRegistry<Material> materialRegistry = new ResourceRegistry<>();
    private final ResourceRegistry<ModelRenderable> modelRenderableRegistry =
            new ResourceRegistry<>();

    private final ResourceRegistry<ViewRenderable> viewRenderableRegistry = new ResourceRegistry<>();

    ResourceRegistry<Texture> getTextureRegistry() {
        return textureRegistry;
    }

    ResourceRegistry<Material> getMaterialRegistry() {
        return materialRegistry;
    }

    ResourceRegistry<ModelRenderable> getModelRenderableRegistry() {
        return modelRenderableRegistry;
    }


    ResourceRegistry<ViewRenderable> getViewRenderableRegistry() {
        return viewRenderableRegistry;
    }

    public long reclaimReleasedResources() {
        long resourcesInUse = 0;
        for (ResourceHolder registry : resourceHolders) {
            resourcesInUse += registry.reclaimReleasedResources();
        }
        return resourcesInUse;
    }

    /**
     * Forcibly deletes all tracked references
     */
    public void destroyAllResources() {
        for (ResourceHolder resourceHolder : resourceHolders) {
            resourceHolder.destroyAllResources();
        }
    }

    public void addResourceHolder(ResourceHolder resource) {
        resourceHolders.add(resource);
    }

    public static ResourceManager getInstance() {
        if (instance == null) {
            instance = new ResourceManager();
        }

        return instance;
    }

    private ResourceManager() {
        addResourceHolder(textureRegistry);
        addResourceHolder(materialRegistry);
        addResourceHolder(modelRenderableRegistry);
        addViewRenderableRegistry();
    }


    private void addViewRenderableRegistry() {
        addResourceHolder(viewRenderableRegistry);
    }
}

File: sceneview/src/main/java/com/google/ar/sceneform/rendering/ResourceManager.kt
package com.google.ar.sceneform.rendering

import com.google.ar.sceneform.resources.ResourceHolder
import com.google.ar.sceneform.resources.ResourceRegistry

/**
 * Minimal resource manager. Maintains mappings from ids to created resources and a task executor
 * dedicated to loading resources asynchronously.
 *
 * @hide
 */
@Suppress("initialization")
class ResourceManager private constructor() {
    private val resourceHolders = ArrayList<ResourceHolder>()
    private val textureRegistry = ResourceRegistry<Texture>()
    private val materialRegistry = ResourceRegistry<Material>()
    private val modelRenderableRegistry = ResourceRegistry<ModelRenderable>()
    private val viewRenderableRegistry = ResourceRegistry<ViewRenderable>()

    init {
        addResourceHolder(textureRegistry)
        addResourceHolder(materialRegistry)
        addResourceHolder(modelRenderableRegistry)
        addViewRenderableRegistry()
    }

    fun getTextureRegistry(): ResourceRegistry<Texture> = textureRegistry

    fun getMaterialRegistry(): ResourceRegistry<Material> = materialRegistry

    fun getModelRenderableRegistry(): ResourceRegistry<ModelRenderable> = modelRenderableRegistry

    fun getViewRenderableRegistry(): ResourceRegistry<ViewRenderable> = viewRenderableRegistry

    fun reclaimReleasedResources(): Long {
        var resourcesInUse = 0L
        for (registry in resourceHolders) {
            resourcesInUse += registry.reclaimReleasedResources()
        }
        return resourcesInUse
    }

    /** Forcibly deletes all tracked references */
    fun destroyAllResources() {
        for (resourceHolder in resourceHolders) {
            resourceHolder.destroyAllResources()
        }
    }

    fun addResourceHolder(resource: ResourceHolder) {
        resourceHolders.add(resource)
    }

    private fun addViewRenderableRegistry() {
        addResourceHolder(viewRenderableRegistry)
    }

    companion object {
        @Volatile
        private var instance: ResourceManager? = null

        @JvmStatic
        fun getInstance(): ResourceManager = instance ?: synchronized(this) {
            instance ?: ResourceManager().also { instance = it }
        }
    }
}
